gremlin-go/driver/connection.go

```diff
@@ -19,50 +19,62 @@ under the License.
 
 package gremlingo
 
+import (
+	"errors"
+	"golang.org/x/text/language"
+)
+
+type connectionState int
+
+const (
+	initialized      connectionState = iota
+	established      connectionState = iota
+	closed           connectionState = iota
+	closedDueToError connectionState = iota
+)
+
 type connection struct {
-	host            string
-	port            int
-	transporterType TransporterType
-	logHandler      *logHandler
-	transporter     transporter
-	protocol        protocol
-	results         map[string]ResultSet
+	logHandler *logHandler
+	protocol   protocol
+	results    map[string]ResultSet
+	state      connectionState
 }
 
-func (connection *connection) close() (err error) {
-	if connection.transporter != nil {
-		err = connection.transporter.Close()
-	}
-	return
+func (connection *connection) errorCallback() {
+	connection.state = closedDueToError
+	_ = connection.protocol.close()
 }
 
-func (connection *connection) connect() error {
-	if connection.transporter != nil {
-		closeErr := connection.transporter.Close()
-		connection.logHandler.logf(Warning, transportCloseFailed, closeErr)
+func (connection *connection) close() (err error) {
+	if connection.state != established {
+		return errors.New("cannot close connection that has already been closed or has not been connected")
 	}
-	connection.protocol = newGremlinServerWSProtocol(connection.logHandler)
-	connection.transporter = getTransportLayer(connection.transporterType, connection.host, connection.port)
-	err := connection.transporter.Connect()
-	if err != nil {
-		return err
+	connection.logHandler.log(Info, closeConnection)
+	if connection.protocol != nil {
+		err = connection.protocol.close()
 	}
-	connection.protocol.connectionMade(connection.transporter)
-	return nil
+	connection.state = closed
+	return
 }
 
 func (connection *connection) write(request *request) (ResultSet, error) {
-	if connection.transporter == nil || connection.transporter.IsClosed() {
-		err := connection.connect()
-		if err != nil {
-			return nil, err
-		}
-	}
-	if connection.results == nil {
-		connection.results = map[string]ResultSet{}
-	}
+	connection.logHandler.log(Info, writeRequest)
+	requestID := request.requestID.String()
+	connection.logHandler.logger.Log(Info, "Creating request with id ", requestID)
```

Comment:
Nit: The request has already been created. Just a minor wording adjustment is probably good here.